tapgarden/re-org_watcher.go

```diff
@@ -0,0 +1,424 @@
+			w.bestHeight = uint32(newBlock)
+
+			for txid := range w.pendingProofs {
+				proofNtfn := w.pendingProofs[txid]
+				firstProof := proofNtfn.firstProof()
+				confs := w.bestHeight - firstProof.BlockHeight
+
+				if confs >= w.cfg.SafeDepth {
+					log.Infof("Anchor TX %v reached %d "+
+						"confirmations, removing it "+
+						"from the re-org watcher",
+						txid, confs)
+
+					// Stop watching the anchor transaction.
+					proofNtfn.cancel()
+
+					delete(w.pendingProofs, txid)
+				}
+			}
+
+		case err := <-blockErr:
+			w.reportErr(fmt.Errorf("unable to receive new block "+
+				"notifications: %w", err))
+			return
+
+		case <-w.Quit:
+			return
+		}
+	}
+}
+
+// WatchProofs adds new proofs to the re-org watcher for their anchor
+// transaction to be watched until it reaches a safe confirmation depth.
+func (w *ReOrgWatcher) WatchProofs(newProofs []*proof.Proof,
+	onProofUpdate ProofUpdateCb) {
+
+	if len(newProofs) == 0 {
+		return
+	}
+
+	select {
+	case w.incomingProofs <- &proofRegistration{
+		proofs:   newProofs,
+		updateCb: onProofUpdate,
+	}:
+	case <-w.Quit:
+	}
+}
+
+// reportErr reports an error to the main server.
+func (w *ReOrgWatcher) reportErr(err error) {
+	select {
+	case w.cfg.ErrChan <- err:
+	case <-w.Quit:
+	}
+}
+
+// waitForConf waits for the anchor transaction of the given proofs to reach a
+// safe confirmation depth.
+func (w *ReOrgWatcher) waitForConf(ctx context.Context,
+	newProofs *proofRegistration) error {
+
+	firstProof := newProofs.proofs[0]
+	tx := firstProof.AnchorTx
+	txHash := tx.TxHash()
+
+	// Do we already have a confirmation watcher for this transaction? Then
+	// we don't need to add another one.
+	existingNtfn, ok := w.pendingProofs[txHash]
+	if ok {
+		existingNtfn.proofsRegistrations = append(
+			existingNtfn.proofsRegistrations, newProofs,
+		)
+
+		return nil
+	}
+
+	reOrgChan := make(chan struct{}, 1)
+	ctxc, cancel := context.WithCancel(ctx)
+
+	// We only register for a single confirmation, which we expect to come
+	// in immediately, since we're only given the proof after one conf.
+	// But we keep the registration open, so we will get notified again if
+	// there is a re-org.
+	confEvent, errChan, err := w.cfg.ChainBridge.RegisterConfirmationsNtfn(
+		ctxc, &txHash, tx.TxOut[0].PkScript, 1, firstProof.BlockHeight,
+		true, reOrgChan,
+	)
+	if err != nil {
+		cancel()
+		return fmt.Errorf("unable to register for conf ntfn: %w", err)
+	}
+
+	// Now that we have created the confirmation watcher, we can add the
+	// new proofs to the pending proofs map.
+	w.pendingProofs[txHash] = &anchorTxNotification{
+		proofsRegistrations: []*proofRegistration{newProofs},
+		cancel:              cancel,
+	}
+
+	w.Wg.Add(1)
+	go func() {
+		defer confEvent.Cancel()
+		defer w.Wg.Done()
+		defer cancel()
+
+		for {
+			select {
+			// If the transaction was confirmed normally, we don't
+			// need to do anything and can stop watching it.
+			case conf, ok := <-confEvent.Confirmed:
```

Comment:
Don't we always expect to his this case? Since it's already confirmed by the time we do the initial registration request.

Reply:
Yes, we should always initially get a confirmation event here. What we're actually interested in are _subsequent_ events in case of a re-org. But I updated some of the log messages so they also make sense for the first confirmation (which we already expect to happen at this point, so shouldn't be `log.Warnf()` for the `Anchor TX ... was already confirmed in block`).